code_api|tool.drcachesim.filter and code_api|tool.drcacheoff.filter broken on AArch64

One of the latest changes to drcachesim seems to have broken AArch64. Unfortunately the AArch64 nightly builds are currently down due to a power failure, but they should be back soon.


    09:56:04.879 132/149 Test #134: code_api|tool.drcachesim.filter .................***Failed  Required regular expression not found.Regex=[^Hello, world!
    09:56:04.879 ---- <application exited with code 0> ----
    09:56:04.879 Cache simulation results:
    09:56:04.879 Core #0 \(1 thread\(s\)\)
    09:56:04.879   L1I stats:
    09:56:04.879     Hits:                         *[0-9,\.]*
    09:56:04.879     Misses:                       *[0-9,\.]*.
    09:56:04.879 .*   Miss rate:                   *[1-9][0-9][,\.]..%
    09:56:04.879   L1D stats:
    09:56:04.879     Hits:                         *[0-9,\.]*
    09:56:04.879     Misses:                       *[0-9,\.]*.
    09:56:04.879 .*   Miss rate:                   *[1-9][0-9][,\.]..%
    09:56:04.879 Core #1 \(0 thread\(s\)\)
    09:56:04.879 Core #2 \(0 thread\(s\)\)
    09:56:04.879 Core #3 \(0 thread\(s\)\)
    09:56:04.879 LL stats:
    09:56:04.879     Hits:                         *[0-9,\.]*.
    09:56:04.879     Misses:                       *[0-9,\.]*..
    09:56:04.879 .*   Local miss rate:             *[1-9][0-9][,\.]..%
    09:56:04.879     Child hits:                   *[0-9,\.]*.
    09:56:04.879     Total miss rate:              *[1-9][0-9][,\.]..%
    09:56:04.879 $
    09:56:04.879 ]  0.20 sec
    09:56:04.879 <Internal Error: Failed to encode instruction: 'lsrv   %x30 $0x06 -> %x30'
    09:56:04.879 >
    09:56:04.879 <Application /var/lib/go-agent/pipelines/dynamorio-verification/build_debug-internal-64/suite/tests/bin/simple_app (27354).  Internal Error: DynamoRIO debug check failure: Not implemented @/var/lib/go-agent/pipelines/dynamorio-verification/dynamorio/core/arch/aarch64/encode.c:172 (0)


    09:56:04.880 135/149 Test #141: code_api|tool.drcacheoff.filter .................***Failed    0.60 sec
    09:56:04.880 Running |/usr/bin/cmake;-E;remove_directory /var/lib/go-agent/pipelines/dynamorio-verification/build_debug-internal-64/suite/tests/drmemtrace.simple_app.26405.1511.dir|
    09:56:04.880 Running cmd |/var/lib/go-agent/pipelines/dynamorio-verification/build_debug-internal-64/bin64/drrun;-s;90;-quiet;-debug;-killpg;-stderr_mask;0xC;-dumpcore_mask;0;-code_api;-t;drcachesim;-offline;-L0_filter;--;/var/lib/go-agent/pipelines/dynamorio-verification/build_debug-internal-64/suite/tests/bin/simple_app|
    09:56:04.880 Running postcmd |/var/lib/go-agent/pipelines/dynamorio-verification/build_debug-internal-64/clients/bin64/drcachesim;-indir;/var/lib/go-agent/pipelines/dynamorio-verification/build_debug-internal-64/suite/tests/drmemtrace.simple_app.27371.4030.dir|
    09:56:04.880 CMake Error at /var/lib/go-agent/pipelines/dynamorio-verification/dynamorio/suite/tests/runmulti.cmake:99 (message):
    09:56:04.880   *** postcmd failed (1): ERROR: Failed to parse module file
    09:56:04.880   /var/lib/go-agent/pipelines/dynamorio-verification/build_debug-internal-64/suite/tests/drmemtrace.simple_app.27371.4030.dir/raw/modules.log
    09:56:04.880 
    09:56:04.880 
    09:56:04.880   ***
    09:56:04.880 
    09:56:04.880 Call Stack (most recent call first):
    09:56:04.880   /var/lib/go-agent/pipelines/dynamorio-verification/dynamorio/suite/tests/runmulti.cmake:111 (process_cmdline)
    09:56:04.880   
    09:56:04.880 
    09:56:04.880 
    09:56:04.880

--- a/core/arch/aarch64/instr_create.h
+++ b/core/arch/aarch64/instr_create.h
@@ -360,8 +360,8 @@
 /**
  * This platform-independent macro creates an instr_t for a logical right shift
  * instruction that does affect the status flags.
- * \param dc  The void * dcontext used to allocate memory for the instr_t.
- * \param d  The opnd_t explicit destination operand for the instruction.
+ * \param dc         The void * dcontext used to allocate memory for the instr_t.
+ * \param d          The opnd_t explicit destination operand for the instruction.
  * \param rm_or_imm  The opnd_t explicit source operand for the instruction.
  */
 /* FIXME i#2440: I'm not sure this is correct.  Use INSTR_CREATE_lsr once available!

--- a/core/arch/aarch64/instr_create.h
+++ b/core/arch/aarch64/instr_create.h
@@ -362,7 +362,7 @@
  * instruction that does affect the status flags.
  * \param dc  The void * dcontext used to allocate memory for the instr_t.
  * \param d  The opnd_t explicit destination operand for the instruction.
- * \param s  The opnd_t explicit source operand for the instruction.
+ * \param rm_or_imm  The opnd_t explicit source operand for the instruction.
  */
 /* FIXME i#2440: I'm not sure this is correct.  Use INSTR_CREATE_lsr once available!
  * Also, what about writing the flags?  Most users don't want to read the flag results,
@@ -371,9 +371,10 @@
  */
 #define XINST_CREATE_slr_s(dc, d, rm_or_imm) \
   (opnd_is_reg(rm_or_imm) ? \
-    instr_create_1dst_2src((dc), OP_lsrv, (d), (d), (rm_or_imm)) : \
-    instr_create_1dst_3src((dc), OP_ubfm, (d), (d), (rm_or_imm), \
-                           (reg_is_32bit(opnd_get_reg(d)) ? OPND_CREATE_INT(31) : OPND_CREATE_INT(63))))
+    instr_create_1dst_2src(dc, OP_lsrv, d, d, rm_or_imm) : \
+    instr_create_1dst_3src(dc, OP_ubfm, d, d, rm_or_imm, \
+                           reg_is_32bit(opnd_get_reg(d)) ? OPND_CREATE_INT(31) : \
+                                                           OPND_CREATE_INT(63)))
 
 /**
  * This platform-independent macro creates an instr_t for a nop instruction.

--- a/core/arch/aarch64/instr_create.h
+++ b/core/arch/aarch64/instr_create.h
@@ -263,7 +263,7 @@
  * precisely where this instruction will be encoded).
  */
 #define XINST_CREATE_jump_cond(dc, pred, t) \
-    (INSTR_PRED(INSTR_CREATE_b((dc), (t)), (pred)))
+    (INSTR_PRED(INSTR_CREATE_bcond((dc), (t)), (pred)))
 
 /**
  * This platform-independent macro creates an instr_t for an addition
@@ -369,8 +369,11 @@
  * they just need to know whether they need to preserve the app's flags, so maybe
  * we can just document that this may not write them.
  */
-#define XINST_CREATE_slr_s(dc, d, s) \
-  instr_create_1dst_2src((dc), OP_lsrv, (d), (d), (s))
+#define XINST_CREATE_slr_s(dc, d, rm_or_imm) \
+  opnd_is_reg(rm_or_imm) ? \
+    instr_create_1dst_2src((dc), OP_lsrv, (d), (d), (rm_or_imm)) : \
+    instr_create_1dst_3src((dc), OP_ubfm, (d), (d), (rm_or_imm), \
+                           (reg_is_32bit(opnd_get_reg(d)) ? OPND_CREATE_INT(31) : OPND_CREATE_INT(63)))
 
 /**
  * This platform-independent macro creates an instr_t for a nop instruction.
@@ -425,7 +428,14 @@
     opnd_add_flags((sht), DR_OPND_IS_SHIFT), (sha))
 #define INSTR_CREATE_b(dc, pc) \
   instr_create_0dst_1src((dc), OP_b, (pc))
-
+/**
+ * This macro creates an instr_t for a conditional branch instruction. The condition
+ * can be set using INSTR_PRED macro.
+ * \param dc The void * dcontext used to allocate memory for the instr_t.
+ * \param pc The opnd_t target operand containing the program counter to jump to.
+ */
+#define INSTR_CREATE_bcond(dc, pc) \
+  instr_create_0dst_1src((dc), OP_bcond, (pc))
 /**
  * This macro creates an instr_t for a BL (branch and link) instruction.
  * \param dc The void * dcontext used to allocate memory for the instr_t.
@@ -445,7 +455,7 @@
 #define INSTR_CREATE_cbz(dc, pc, reg) \
   instr_create_0dst_2src((dc), OP_cbz, (pc), (reg))
 #define INSTR_CREATE_cmp(dc, rn, rm_or_imm) \
-  instr_create_1dst_2src(dc, OP_subs, OPND_CREATE_ZR(rn), rn, rm_or_imm)
+  INSTR_CREATE_subs(dc, OPND_CREATE_ZR(rn), rn, rm_or_imm)
 #define INSTR_CREATE_ldp(dc, rt1, rt2, mem) \
   instr_create_2dst_1src(dc, OP_ldp, rt1, rt2, mem)
 #define INSTR_CREATE_ldr(dc, Rd, mem) \

--- a/core/arch/aarch64/instr_create.h
+++ b/core/arch/aarch64/instr_create.h
@@ -370,10 +370,10 @@
  * we can just document that this may not write them.
  */
 #define XINST_CREATE_slr_s(dc, d, rm_or_imm) \
-  opnd_is_reg(rm_or_imm) ? \
+  (opnd_is_reg(rm_or_imm) ? \
     instr_create_1dst_2src((dc), OP_lsrv, (d), (d), (rm_or_imm)) : \
     instr_create_1dst_3src((dc), OP_ubfm, (d), (d), (rm_or_imm), \
-                           (reg_is_32bit(opnd_get_reg(d)) ? OPND_CREATE_INT(31) : OPND_CREATE_INT(63)))
+                           (reg_is_32bit(opnd_get_reg(d)) ? OPND_CREATE_INT(31) : OPND_CREATE_INT(63))))
 
 /**
  * This platform-independent macro creates an instr_t for a nop instruction.

